mbj4668/pyang
# Fix issue for #628 unexpected keyword can be used within the "augment…

**Solution:**

Add semantic check in the function of v_expand_2_augment.

1. A new function of keyword_is_expected is introduced, it is used to check whether the nodes defined in the substatements that follow the "augment" statement are expected.
2. A new error code of UNEXPECTED_KEYWORD_AUGMENT is introduced.

**Test procedure:**

The following 5 test cases are used to verify the current fix:

**Case 1: target node is case, the "action" and "notification" statements used within the "augment" statement**

```
module mod1 {
  yang-version 1.1;
  namespace "urn:mod1";
  prefix m1;

  container interfaces {
    list ifEntry {
      key "ifIndex";
      leaf ifIndex {
        type uint32;
      }

      choice nodechoice {
        case nodecase {
          leaf name {
            type string;
          }
        }
      }
    }
  }

  augment "/interfaces/ifEntry/nodechoice/nodecase" {//target node is case, it is not a container or list node.
    leaf result {
      type string;
    }
    notification nodenotification {//error: unexpected keyword 
      leaf reset-at {
        type string;
      }
    }

    action nodeaction {//error: unexpected keyword 
      input {
        leaf start {
          type string;
        }
      }
      output {
        leaf finish {
          type string;
        }
      }
    }
  }
}

```
**Case 2: target node is list, the "case" statement used within the "augment" statement.**

```
module mod2 {
  yang-version 1.1;
  namespace "urn:mod2";
  prefix m2;

  container interfaces {
    list ifEntry {
      key "ifIndex";
      leaf ifIndex {
        type uint32;
      }
    }
  }

  augment "/interfaces/ifEntry" {//target node is list, it is not a choice node.
    leaf result {
      type string;
    }

    case nodecase {//error: unexpected keyword 
      leaf test {
        type string;
      }
    }
  }
}


```
**Case 3: target node is choice, the "action" and "notification" statements used within the "augment" statement**

```
module mod3 {
  yang-version 1.1;
  namespace "urn:mod3";
  prefix m3;

  container interfaces {
    list ifEntry {
      key "ifIndex";
      leaf ifIndex {
        type uint32;
      }

      choice nodechoice {
        case nodecase {
          leaf name {
            type string;
          }
        }
      }
    }
  }

  augment "/interfaces/ifEntry/nodechoice" {//target node is choice, it is not a container or list node.
    leaf result {
      type string;
    }
    
	notification resetnotify {//error: unexpected keyword 
      leaf reset-at {
        type string;
      }
    }
	
    action resetact {//error: unexpected keyword 
      input {
        leaf start {
          type string;
        }
      }
      output {
        leaf finish {
          type string;
        }
      }
    }
  }
}

```
**Case 4: target node is output, the "action" and "notification" statements used within the "augment" statement**
```
module mod4 {
  yang-version 1.1;
  namespace "urn:mod4";
  prefix m4;

  container interfaces {
    list ifEntry {
      key "ifIndex";
      leaf ifIndex {
        type uint32;
      }

      action nodeaction {
        input {
          leaf start {
            type string;
          }
        }
        output {
          leaf finish {
            type string;
          }
        }
      }
    }
  }

  augment "/interfaces/ifEntry/nodeaction/output" {//target node is output, it is not a container or list node.
    leaf result {
      type string;
    }
    notification resetnotify {//error: unexpected keyword 
      leaf reset-at {
        type string;
      }
    }

    action resetact {//error: unexpected keyword 
      input {
        leaf start {
          type string;
        }
      }
      output {
        leaf finish {
          type string;
        }
      }
    }
  }
}

```
**Case 5: target node is notification, the "action" and "notification" statements used within the "augment" statement**
```
module mod5 {
  yang-version 1.1;
  namespace "urn:mod5";
  prefix m5;

  container interfaces {
    list ifEntry {
      key "ifIndex";
      leaf ifIndex {
        type uint32;
      }

      action resetact {
        input {
          leaf start {
            type string;
          }
        }
        output {
          leaf finish {
            type string;
          }
        }
      }
      
	  notification nodenotification {
          leaf reset-at2 {
            type string;
            mandatory true;
          }
        }
    }
  }

  augment "/interfaces/ifEntry/nodenotification" {//target node is a notification, it is not a container or list node.
    leaf result {
      type string;
    }

    notification resetnotify {//error: unexpected keyword
      leaf reset-at {
        type string;
      }
    }

    action resetact {//error: unexpected keyword
      input {
        leaf start {
          type string;
        }
      }
      output {
        leaf finish {
          type string;
        }
      }
    }  
  }
}
```
